xarray/namedarray/core.py

```diff
@@ -847,5 +855,108 @@ def _to_dense(self) -> Self:
         else:
             raise TypeError("self.data is not a sparse array")
 
+    def transpose(
+        self,
+        *dims: Hashable | ellipsis,
+        missing_dims: ErrorOptionsWithWarn = "raise",
+    ) -> Self:
+        """Return a new object with transposed dimensions.
+
+        Parameters
+        ----------
+        *dims : Hashable, optional
+            By default, reverse the dimensions. Otherwise, reorder the
+            dimensions to this order.
+        missing_dims : {"raise", "warn", "ignore"}, default: "raise"
+            What to do if dimensions that should be selected from are not present in the
+            NamedArray:
+            - "raise": raise an exception
+            - "warn": raise a warning, and ignore the missing dimensions
+            - "ignore": ignore the missing dimensions
+
+        Returns
+        -------
+        transposed : NamedArray
+            The returned object has transposed data and dimensions with the
+            same attributes as the original.
+
+        Notes
+        -----
+        This operation returns a view of this variable's data. It is
+        lazy for dask-backed Variables but not for numpy-backed Variables.
+
+        See Also
+        --------
+        numpy.transpose
+        """
+        from xarray.core.indexing import as_indexable
+
+        if len(dims) == 0:
+            dims = self.dims[::-1]
+        else:
+            dims = tuple(infix_dims(dims, self.dims, missing_dims))
+
+        if len(dims) < 2 or dims == self.dims:
+            # no need to transpose if only one dimension
+            # or dims are in same order
+            return self.copy(deep=False)
+
+        axes = self.get_axis_num(dims)
+        data = as_indexable(self._data).transpose(axes)
+        return self._replace(dims=dims, data=data)
+
+    @property
+    def T(self) -> Self:
+        return self.transpose()
+
+    def set_dims(self, dims: _DimsLike, shape: _ShapeLike = None) -> Self:
```

Comment:
We wanted to split this into `broadcast_to` and `expand_dims` in the [design doc](https://github.com/pydata/xarray/blob/main/design_notes/named_array_design_doc.md), so perhaps a separate PR?

Reply:
yes, i initially thought it would be easy to port this code, and then split it into `broadcast_to` and `expand_dims` in a separate pull request

Reply:
Either way...

Reply:
@dcherian, i went ahead and split `set_dims()` into `.expand_dims()` and `broadcast_to()` as part of this pr